// minecraft-server-util/dist/scanLAN.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const dgram_1 = __importDefault(require("dgram"));
const util_1 = require("util");
const minecraft_motd_util_1 = require("minecraft-motd-util");
const pattern = /\[MOTD\](.+)?(?=\[\/MOTD\])\[\/MOTD\]\[AD\](\d+)\[\/AD\]/;
const decoder = new util_1.TextDecoder('utf-8');
function applyDefaultOptions(options) {
    // Apply the provided options on the default options
    return Object.assign({
        scanTime: 1000 * 5
    }, options);
}
/**
 * Scans the local area network for any Minecraft worlds.
 * @param {ScanLANOptions} [options] The options to use when scanning LAN
 * @returns {Promise<ScanLANResponse>} The response of the scan
 * @async
 */
function scanLAN(options) {
    return __awaiter(this, void 0, void 0, function* () {
        // Applies the provided options on top of the default options
        const opts = applyDefaultOptions(options);
        // Assert that the arguments are the correct type and format
        (0, assert_1.default)(typeof options === 'object' || typeof options === 'undefined', `Expected 'options' to be an object or undefined, got ${typeof options}`);
        (0, assert_1.default)(typeof opts === 'object', `Expected 'options' to be an object, got ${typeof opts}`);
        (0, assert_1.default)(typeof opts.scanTime === 'number', `Expected 'options.scanTime' to be a number, got ${typeof opts.scanTime}`);
        (0, assert_1.default)(opts.scanTime >= 1500, `Expected 'options.scanTime' to be greater than or equal to 1500, got ${opts.scanTime}`);
        // Create a list of servers that the socket will append to
        const servers = [];
        // Create a new UDP socket and listen for messages
        const socket = dgram_1.default.createSocket('udp4');
        // Wait for messages being broadcased
        socket.on('message', (message, info) => {
            const text = decoder.decode(message);
            // Ensure that the text sent to the scan port matches the "Open to LAN" format
            if (!pattern.test(text))
                return;
            const match = text.match(pattern);
            if (!match)
                return;
            // Parse the port out of the matched text
            const port = parseInt(match[2]);
            if (isNaN(port))
                return;
            const server = servers.find((server) => server.host === info.address && server.port === port);
            if (server)
                return;
            const description = (0, minecraft_motd_util_1.parse)(match[1]);
            const motd = {
                raw: (0, minecraft_motd_util_1.format)(description),
                clean: (0, minecraft_motd_util_1.clean)(description),
                html: (0, minecraft_motd_util_1.toHTML)(description)
            };
            servers.push({
                host: info.address,
                port,
                motd
            });
        });
        // Bind to the 4445 port which is used for receiving and broadcasting "Open to LAN" worlds
        socket.bind(4445, () => {
            socket.addMembership('224.0.2.60');
        });
        // Return the timeout promise that will resolve when the scan time is up
        return new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                resolve({ servers });
                socket.close();
            }, opts.scanTime);
            socket.on('error', (error) => {
                socket.close();
                clearTimeout(timeout);
                reject(error);
            });
        });
    });
}
exports.default = scanLAN;
