// minecraft-server-util/dist/util/formatResultFE01FA.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const minecraft_motd_util_1 = require("minecraft-motd-util");
/**
 * Formats the raw response from a status request into a more useable format
 * @param {string} host The host of the server
 * @param {number} port The port of the server
 * @param {SRVRecord | null} srvRecord The SRV lookup data
 * @param {number} protocolVersion The protocol version returned from the server
 * @param {string} version The version string of the server
 * @param {string} motd The MOTD of the server
 * @param {number} onlinePlayers The amount of players in the server
 * @param {number} maxPlayers The maximum amount of players in the server
 * @returns {StatusResponse} The formatted result
 */
function formatResultFE01FA(host, port, srvRecord, protocolVersion, version, motd, onlinePlayers, maxPlayers, roundTripLatency) {
    (0, assert_1.default)(typeof host === 'string', 'Expected host to be a string, got ' + typeof host);
    (0, assert_1.default)(host.length > 0, 'Expected host.length > 0, got ' + host.length);
    (0, assert_1.default)(typeof port === 'number', 'Expected port to be a number, got ' + typeof port);
    (0, assert_1.default)(Number.isInteger(port), 'Expected port to be an integer, got ' + port);
    (0, assert_1.default)(port > 0, 'Expected port > 0, got ' + port);
    (0, assert_1.default)(port < 65536, 'Expected port < 65536, got ' + port);
    (0, assert_1.default)(typeof protocolVersion === 'number', 'Expected protocolVersion to be a number, got ' + typeof protocolVersion);
    (0, assert_1.default)(Number.isInteger(protocolVersion), 'Expected protocolVersion to be an integer, got ' + protocolVersion);
    (0, assert_1.default)(protocolVersion >= 0, 'Expected protocolVersion >= 0, got ' + protocolVersion);
    (0, assert_1.default)(typeof version === 'string', 'Expected version to be a string, got ' + typeof version);
    (0, assert_1.default)(version.length > 0, 'Expected version.length > 0, got ' + version.length);
    (0, assert_1.default)(typeof onlinePlayers === 'number', 'Expected onlinePlayers to be a number, got ' + typeof onlinePlayers);
    (0, assert_1.default)(Number.isInteger(onlinePlayers), 'Expected onlinePlayers to be an integer, got ' + onlinePlayers);
    (0, assert_1.default)(typeof maxPlayers === 'number', 'Expected maxPlayers to be a number, got ' + typeof maxPlayers);
    (0, assert_1.default)(Number.isInteger(maxPlayers), 'Expected maxPlayers to be an integer, got ' + maxPlayers);
    const description = (0, minecraft_motd_util_1.parse)(motd);
    return {
        host,
        port,
        srvRecord,
        version,
        protocolVersion,
        onlinePlayers,
        maxPlayers,
        samplePlayers: null,
        motd: description ? {
            raw: (0, minecraft_motd_util_1.format)(description),
            clean: (0, minecraft_motd_util_1.clean)(description),
            html: (0, minecraft_motd_util_1.toHTML)(description)
        } : null,
        favicon: null,
        modInfo: null,
        rawResponse: null,
        roundTripLatency
    };
}
exports.default = formatResultFE01FA;
