// minecraft-server-util/dist/util/formatResult.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const minecraft_motd_util_1 = require("minecraft-motd-util");
const Favicon_1 = __importDefault(require("../structure/Favicon"));
/**
 * Formats the raw response from a status request into a more useable format
 * @param {string} host The host of the server
 * @param {number} port The port of the server
 * @param {SRVRecord | null} srvRecord The SRV lookup data
 * @param {RawStatusResponse} result The raw JSON data returned from the server
 * @returns {StatusResponse} The formatted result
 */
function formatResult(host, port, srvRecord, result, roundTripLatency) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    (0, assert_1.default)(typeof host === 'string', 'Expected host to be a string, got ' + typeof host);
    (0, assert_1.default)(host.length > 0, 'Expected host.length > 0, got ' + host.length);
    (0, assert_1.default)(typeof port === 'number', 'Expected port to be a number, got ' + typeof port);
    (0, assert_1.default)(Number.isInteger(port), 'Expected integer, got ' + port);
    (0, assert_1.default)(port > 0, 'Expected port > 0, got ' + port);
    (0, assert_1.default)(port < 65536, 'Expected port < 65536, got ' + port);
    (0, assert_1.default)(typeof result === 'object', 'Expected object, got ' + (typeof result));
    (0, assert_1.default)(typeof roundTripLatency === 'number', 'Expected roundTripLatency to be a number, got ' + typeof roundTripLatency);
    (0, assert_1.default)(Number.isInteger(roundTripLatency), 'Expected roundTripLatency to be an integer, got ' + roundTripLatency);
    (0, assert_1.default)(roundTripLatency >= 0, 'Expected roundTripLatency >= 0, got ' + roundTripLatency);
    const version = (_b = (_a = result === null || result === void 0 ? void 0 : result.version) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : null;
    const protocolVersion = (_d = (_c = result === null || result === void 0 ? void 0 : result.version) === null || _c === void 0 ? void 0 : _c.protocol) !== null && _d !== void 0 ? _d : null;
    const onlinePlayers = (_f = (_e = result === null || result === void 0 ? void 0 : result.players) === null || _e === void 0 ? void 0 : _e.online) !== null && _f !== void 0 ? _f : null;
    const maxPlayers = (_h = (_g = result === null || result === void 0 ? void 0 : result.players) === null || _g === void 0 ? void 0 : _g.max) !== null && _h !== void 0 ? _h : null;
    const samplePlayers = (_k = (_j = result === null || result === void 0 ? void 0 : result.players) === null || _j === void 0 ? void 0 : _j.sample) !== null && _k !== void 0 ? _k : null;
    const description = typeof result.description !== 'undefined' ? (0, minecraft_motd_util_1.parse)(result.description) : null;
    const favicon = (_l = result === null || result === void 0 ? void 0 : result.favicon) !== null && _l !== void 0 ? _l : null;
    const modInfo = (_m = result === null || result === void 0 ? void 0 : result.modinfo) !== null && _m !== void 0 ? _m : null;
    const response = {
        host,
        port,
        srvRecord,
        version,
        protocolVersion,
        onlinePlayers,
        maxPlayers,
        samplePlayers,
        motd: description ? {
            raw: (0, minecraft_motd_util_1.format)(description),
            clean: (0, minecraft_motd_util_1.clean)(description),
            html: (0, minecraft_motd_util_1.toHTML)(description)
        } : null,
        favicon: favicon ? new Favicon_1.default(favicon) : null,
        modInfo,
        roundTripLatency
    };
    Object.defineProperty(response, 'rawResponse', {
        value: result,
        enumerable: false
    });
    return response;
}
exports.default = formatResult;
